File: educational-core/src/com/jetbrains/edu/coursecreator/actions/placeholder/CCEditAnswerPlaceholder.java
package com.jetbrains.edu.coursecreator.actions.placeholder;

import com.intellij.openapi.actionSystem.Presentation;
import com.intellij.openapi.project.Project;
import com.jetbrains.edu.coursecreator.actions.placeholder.CCCreateAnswerPlaceholderDialog.DependencyInfo;
import com.jetbrains.edu.learning.EduState;
import com.jetbrains.edu.learning.courseFormat.AnswerPlaceholder;
import com.jetbrains.edu.learning.courseFormat.AnswerPlaceholderDependency;
import com.jetbrains.edu.learning.messages.EduCoreBundle;
import com.jetbrains.edu.learning.yaml.YamlFormatSynchronizer;
import org.jetbrains.annotations.NotNull;

public class CCEditAnswerPlaceholder extends CCAnswerPlaceholderAction {

  public CCEditAnswerPlaceholder() {
    super(() -> EduCoreBundle.message("label.edit"), () -> EduCoreBundle.message("action.edit.answer.placeholder.description"));
  }

  @Override
  protected void performAnswerPlaceholderAction(@NotNull Project project, @NotNull EduState state) {
    AnswerPlaceholder answerPlaceholder = state.getAnswerPlaceholder();
    if (answerPlaceholder == null) {
      return;
    }
    performEditPlaceholder(project, answerPlaceholder);
  }

  public void performEditPlaceholder(@NotNull Project project, @NotNull AnswerPlaceholder answerPlaceholder) {
    CCCreateAnswerPlaceholderDialog dlg = createDialog(project, answerPlaceholder);

    if (dlg.showAndGet()) {
      final String answerPlaceholderText = dlg.getPlaceholderText();
      answerPlaceholder.setPlaceholderText(answerPlaceholderText);
      final DependencyInfo dependencyInfo = dlg.getDependencyInfo();
      if (dependencyInfo != null) {
        answerPlaceholder.setPlaceholderDependency(
          AnswerPlaceholderDependency.create(answerPlaceholder, dependencyInfo.getDependencyPath(), dependencyInfo.isVisible()));
      }
      YamlFormatSynchronizer.saveItem(answerPlaceholder.getTaskFile().getTask());
    }
  }

  protected CCCreateAnswerPlaceholderDialog createDialog(@NotNull Project project, @NotNull AnswerPlaceholder answerPlaceholder) {
    return new CCCreateAnswerPlaceholderDialog(project, true, answerPlaceholder);
  }

  @Override
  protected void updatePresentation(@NotNull EduState eduState, @NotNull Presentation presentation) {
    presentation.setEnabledAndVisible(eduState.getAnswerPlaceholder() != null);
  }
}

File: educational-core/src/com/jetbrains/edu/coursecreator/actions/placeholder/CCEditAnswerPlaceholder.kt
package com.jetbrains.edu.coursecreator.actions.placeholder

import com.intellij.openapi.actionSystem.Presentation
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.fileEditor.FileEditorManager
import com.jetbrains.edu.learning.EduState
import com.jetbrains.edu.learning.courseDir
import com.jetbrains.edu.learning.courseFormat.ext.getDir
import com.jetbrains.edu.learning.yaml.configFileName
import org.jetbrains.annotations.NonNls

open class CCEditAnswerPlaceholder : CCAnswerPlaceholderAction() {

  override fun performAnswerPlaceholderAction(state: EduState) {
    val answerPlaceholder = state.answerPlaceholder ?: return
    val task = answerPlaceholder.taskFile.task
    val configFileName = task.configFileName
    val taskDir = task.getDir(state.project.courseDir)
    if (taskDir == null) {
      LOG.error("Failed to find task directory")
      return
    }
    val configFile = taskDir.findChild(configFileName)
    if (configFile == null) {
      LOG.error("Failed to find task config file")
      return
    }
    FileEditorManager.getInstance(state.project).openFile(configFile, true)
  }

  override fun updatePresentation(eduState: EduState, presentation: Presentation) {
    presentation.isEnabledAndVisible = eduState.answerPlaceholder != null
  }

  companion object {
    private val LOG = Logger.getInstance(CCEditAnswerPlaceholder::class.java)

    @NonNls
    const val ACTION_ID = "Educational.Educator.EditAnswerPlaceholder"
  }
}